cache/metrics.go

```diff
@@ -0,0 +1,92 @@
+package cache
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	instrumentationName      = "github.com/moby/buildkit/cache"
+	metricCacheRecords       = "cache.records.count"
+	metricCachePruneDuration = "cache.prune.duration"
+)
+
+type metrics struct {
+	CacheRecords       metric.Int64ObservableGauge
+	CachePruneDuration metric.Int64Histogram
+	meter              metric.Meter
+	regs               []metric.Registration
+}
+
+func newMetrics(cm *cacheManager, mp metric.MeterProvider) (*metrics, error) {
+	m := &metrics{}
+
+	var err error
+	m.meter = mp.Meter(instrumentationName)
+
+	m.CacheRecords, err = m.meter.Int64ObservableGauge(metricCacheRecords,
+		metric.WithDescription("Number of cache records."),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	m.CachePruneDuration, err = m.meter.Int64Histogram(metricCachePruneDuration,
+		metric.WithDescription("Measures the duration of cache prune operations."),
+		metric.WithUnit("ms"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	reg, err := m.meter.RegisterCallback(cm.collectMetrics, m.CacheRecords)
+	if err != nil {
+		return nil, err
+	}
+	m.regs = append(m.regs, reg)
+
+	return m, nil
+}
+
+func (m *metrics) MeasurePrune() (record func(ctx context.Context)) {
+	if m == nil {
+		return func(_ context.Context) {}
+	}
+
+	start := time.Now()
+	return func(ctx context.Context) {
+		dur := int64(time.Since(start) / time.Millisecond)
+		m.CachePruneDuration.Record(ctx, dur)
+	}
+}
+
+func (m *metrics) Close() error {
+	if m == nil {
+		return nil
+	}
+
+	for _, reg := range m.regs {
+		_ = reg.Unregister()
+	}
+	return nil
+}
+
+type cacheStats struct {
+	NumRecords int64
+}
+
+func (cm *cacheManager) readStats() (stats cacheStats) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	stats.NumRecords = int64(len(cm.records))
+	return stats
+}
+
+func (cm *cacheManager) collectMetrics(ctx context.Context, o metric.Observer) error {
```

Comment:
When does this get called?

Reply:
This gets called automatically during a metrics collection interval (defined by the reader). So for the periodic reader, every 60 seconds by default. For the prometheus reader, whenever `/metrics` is called.

See the invocation of `RegisterCallback` earlier in this file and the observable gauge.

Reply:
> So for the periodic reader, every 60 seconds by default. For the prometheus reader, whenever /metrics is called.

q: Does this mean that the calls are duplicated because there are 2 readers?

Reply:
Just checked this. Short answer, yes.

The call happens for each reader so it'll happen each time `/metrics` is hit and also every 60 seconds.

Alternatively, if the call never happens because the metrics are never checked, it happens zero times. So if the otel collector isn't configured and the prometheus endpoint doesn't exist or isn't invoked.